### PR TITLE
[12.0][ADD] project_task_project_required

### PR DESCRIPTION
**Pull request by alexey-pelykh**
This module allows to configure task as having project as a mandatory field.

### CODE REVIEW

File: project_task_project_required/models/res_config_settings.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    is_project_task_project_required = fields.Boolean(
+        string='Require Projects on Tasks',
+        related='company_id.is_project_task_project_required',
+        readonly=False,
```
Comment (tarteo) [inline project_task_project_required/models/res_config_settings.py:13]: Isn't this the default for related fields?

**Review by tarteo — APPROVED** [review]
Very nice!

File: project_task_project_required/models/res_config_settings.py (hunk 1 above)
Comment (alexey-pelykh) [reply project_task_project_required/models/res_config_settings.py:13]: Sinve v12 default is `readonly=True`, see https://github.com/odoo/odoo/blob/897feb553e6ef75af9434cb77436b26abde3a761/odoo/fields.py#L427

Comment (tarteo) [reply project_task_project_required/models/res_config_settings.py:13]: I see, thanks :+1:

File: project_task_project_required/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    'summary': """
+        Set project on task as a mandatory field""",
+    'version': '12.0.1.0.0',
+    'license': 'AGPL-3',
+    'author':
+        'Brainbean Apps, '
+        'Odoo Community Association (OCA)',
+    'website': 'https://github.com/OCA/project',
+    'depends': [
+        'project',
```
Comment (Saran440) [inline project_task_project_required/__manifest__.py:15]:
```suggestion
        'project',
        'hr_timesheet',
```
Error when I not install hr_timesheet module.

File: project_task_project_required/models/project_task.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    is_project_required = fields.Boolean(
+        string='Is Project Required',
+        related='company_id.is_project_task_project_required',
+    )
+
+    @api.constrains('project_id')
+    def _check_project_id(self):
+        for task in self:
+            if task.is_project_required and not task.project_id:
```
Comment (Saran440) [inline project_task_project_required/models/project_task.py:19]: I don't know how can it happen in case?

**Review by Saran440 — COMMENTED** [review]
Functional Test  :+1: 
and Code Review

File: project_task_project_required/models/project_task.py (hunk 3 above)
Comment (alexey-pelykh) [reply project_task_project_required/models/project_task.py:19]: Odoo allows creating tasks and unsetting project_id after task has been created, so this check is the basically very essence of the module.

File: project_task_project_required/__manifest__.py (hunk 2 above)
Comment (alexey-pelykh) [reply project_task_project_required/__manifest__.py:15]: Fixed origin of the issue

File: project_task_project_required/models/project_task.py (hunk 3 above)
Comment (Saran440) [reply project_task_project_required/models/project_task.py:19]: My steps test
1.Go to Project -> All Task -> Create Task (don't depend project_id)
2.Go to Project -> Configuration -> check `Require Projects on Tasks`
3.Back to Task -> Project will required field.

Project is required field already and why are you want to check this function? or any case that used this function.

Comment (alexey-pelykh) [reply project_task_project_required/models/project_task.py:19]: That's marking UI field as required, it has nothing to do with function that ensures data integrity if I want to create task without project programmatically

**Comment by alexey-pelykh** [comment]
Right now I have no idea why this build started to fail due to field that is not referenced anywhere:
```
Field 'allowed_user_ids' used in attributes must be present in view but is missing:
- 'allowed_user_ids' in domain="[('id', 'in', allowed_user_ids)]"
```

**Comment by alexey-pelykh** [comment]
For me it's being green: https://travis-ci.org/brainbeanapps/project/builds/578649199
In OCA it fails: https://travis-ci.org/OCA/project/builds/578649217

**Comment by alexey-pelykh** [comment]
@OCA/project-service-maintainers I guess this needs some high-level attention, if possible. 
```
<field name="allowed_user_ids" invisible="1"/>
```
adding this fixed OCA Travis, but breaks Runbot and our Travis. Not having that line (which seems really correct to me) is perfectly fine for Runbot and our Travis, but breaks OCA Travis. Any ideas?

**Review by Saran440 — APPROVED** [review]

**Comment by NL66278** [comment]
@alexey-pelykh As far as I can see the field is defined here: project_hr/models/project_task.py:34. So your module should depend on project_hr.

**Comment by alexey-pelykh** [comment]
@NL66278 thanks for a hit, I have no idea how I've missed it, I've done rgrep on all fresh OCA repos and rebased. Yet I'm not using that field, and this module has no relation at all with `project_hr` - basically I'll force to install unneeded module in this case. Could it be a not quite correct approach in `project_hr` to set domain on field instead of setting it on views?

**Comment by alexey-pelykh** [comment]
@pedrobaeza that module is related to you, I believe. Any ideas on that?

**Comment by rafaelbn** [comment]
Hi, thanks for this contribution and fixing

I'm not sure if [[FIX] project_hr: domain on field breaks other modules](https://github.com/OCA/project/pull/570/commits/3f4d22200659848b2c364ce6b587f519fbd70305) should be fixed in a separate PR

@pedrobaeza is out of office

**Comment by alexey-pelykh** [comment]
@rafaelbn easily workaroundable https://github.com/OCA/project/pull/573

**Comment by rafaelbn** [comment]
Please @alexey-pelykh https://github.com/OCA/project/pull/573 is merged. Rebase

**Comment by alexey-pelykh** [comment]
@rafaelbn all good

**Review by rafaelbn — APPROVED** [review]

**Pull request merged**